Prevent Imbalanced transaction in Journal

lomamech opened the pull request:
Prevent Imbalanced transaction in Journal  …
- Adding function validate transaction who check if debit is equal to  credit
- Restructured catching error in controller of journal

closes #1256

lomamech committed:
Prevent Imbalanced transaction in Journal
- Adding function validate transaction who check if debit is equal to
  credit
- Restructured catching error in controller of journal

jniles commented on client/src/js/services/grid/TransactionService.js:
```diff
@@ -295,6 +295,26 @@ function TransactionService($timeout, util, uiGridConstants, bhConstants, Notify
+      credit += Number(journal.credit_equiv);
+    });
+
+    return debit !== credit ? true : false;
```
In the future, you can just `return debit !== credit`.  It's probably easier to read.

jniles commented on client/src/js/services/grid/TransactionService.js:
```diff
@@ -532,6 +552,12 @@ function TransactionService($timeout, util, uiGridConstants, bhConstants, Notify
    * This function saves all transactions by
    */
   Transactions.prototype.save = function save() {
+    var clientErrors = validateTransaction(this._entity);
```
A better way than returning (and rejecting) a boolean would be to return the error code.  So, instead of returning `true` or `false`, return the string `POSTING_JOURNAL.ERRORS.UNBALANCED_TRANSACTIONS`. 

That way, anyone else who makes an error condition (like missing dates, missing values, etc) can simply add to the validate function.

jniles replied:
The best way to do this would be to add a validation key at the top of this service, like:
```js
var ERR_IMBALANCED_TXN = 'POSTING_JOURNAL.ERRORS.UNBALANCED_TRANSACTIONS';

// later in validateTransaction()
var error;

if (debit !== credit) {
  error = ERR_UNBALANCED_TXN;
}

// another condition

// another condition ..

return error;
```

jniles commented on client/src/js/services/grid/TransactionService.js:
```diff
@@ -295,6 +295,26 @@ function TransactionService($timeout, util, uiGridConstants, bhConstants, Notify
+   * This method is used to check if a transaction is balanced by their record
+   */
+  function validateTransaction(entity) {
+    var dataEntity = entity.data.data;
```
A better name for this is `transaction`, since we are editing a transaction and the function is called `validateTransaction()`.

jniles commented on client/src/js/services/grid/TransactionService.js:
```diff
@@ -295,6 +295,26 @@ function TransactionService($timeout, util, uiGridConstants, bhConstants, Notify
+    var debit = 0,
+      credit = 0;
+    
+    dataEntity.forEach(function (journal) {
```
A transaction is made of rows, not journals (a journal contains transactions), so a better name for this variable is `row`.  Or `line`.  Or `item`.  But `journal` is confusing.

jniles reviewed (changes requested):
This is a good start.  If we want to add more checks later, we will need to move all the logic into `validateTransacation()`.  Why don't you do that from the beginning?

lomamech committed:
Review for prevent imbalanced transaction
-Using the best name for function who check the balance
-Using not confusing key word
-Return the String for the function transaction

jniles commented on client/src/js/services/grid/TransactionService.js:
```diff
@@ -295,6 +295,35 @@ function TransactionService($timeout, util, uiGridConstants, bhConstants, Notify
+   * @description
+   * This method is used to check if a transaction is balanced by their record
+   */
+  function transaction(entity) {
```
@lomamech, the name `validateTransaction()` was fine.  It's `dataEntity` that should be called `transaction`.  Because it _is_ a transaction, right?

lomamech committed:
Improve the label of function and variable

jniles commented on client/src/js/services/grid/TransactionService.js:
```diff
@@ -295,6 +295,35 @@ function TransactionService($timeout, util, uiGridConstants, bhConstants, Notify
+   * This method is used to check if a transaction is balanced by their record
+   */
+  function validateTransaction(entity) {
+    var transaction = entity.data.data;
```
:+1:  This is much clearer!
